File: Java/1646. Get Maximum in Generated Array.java
class Solution {
    public int getMaximumGenerated(int n) {
        int[] nums = new int[101];
        nums[0] = 0;
        nums[1] = 1;
        int max = (n == 0) ? 0 : 1;
        for (int i = 2; i <= n; i++) {
            if (i % 2 == 0) {
                nums[i] = nums[i / 2];
            } else {
                nums[i] = nums[i / 2] + nums[i / 2 + 1];
            }
            max = (nums[i] > max) ? nums[i] : max;
        }
        return max;
    }
}

File: 2021/1646. Get Maximum in Generated Array.cpp
class Solution {
public:
    int getMaximumGenerated(int n) {
        if (n == 0 || n == 1) return n;
        int ans = 1;
        int dp[n + 1];
        dp[0] = 0;
        dp[1] = 1;
        bool by2 = true;
        for (int i = 2; i <= n; i++) {
            dp[i] = by2 ? dp[i / 2] : dp[(i - 1) / 2] + dp[(i + 1) / 2];
            ans = max(ans, dp[i]);
            by2 = !by2;
        }
        return ans;
    }
};